### ISSUE TITLE
No Way to 'Invoke' User-Defined Functions

### ISSUE THREAD

**Issue by elias-winberg** [amir20/phantomjs-node#487]
As of right now, there is no way to invoke user-defined functions. I suggest that we add two methods to `Page` and `Phantom`:

`#.invokeAsyncFunction`:

Can be used to invoke an asynchronous function (a function that takes a callback function).

```js
page.property('exists', function (selector, callback) {
	var element = this.evaluate(function (selector) {
		return document.querySelector(selector);
	}, selector);
	callback(element !== null);
}).then(function () {
	return page.invokeAsyncFunction('exists', 'a');
}).then(function (exists) {
	assert(exists);
});
```

```js
page.invokeAsyncFunction('open', 'http://stackoverflow.com/').then(function (status) {
	assert(status === 'success');
});
```

`#.invokeFunction`:

Can be used to invoke a function.

```js
page.property('exists', function (selector) {
	var element = this.evaluate(function (selector) {
		return document.querySelector(selector);
	}, selector);
	return element !== null;
}).then(function () {
	return page.invokeFunction('exists', 'a');
}).then(function (exists) {
	assert(exists);
});
```

```js
page.invokeFunction('evaluate', function () {
	return document.title;
}).then(function (title) {
	console.log(title);
});
```

We could also bind the builtin/predefined functions to these functions:

```js
// src/page.js

const asyncMethods = [
	'includeJs',
	'open'
];

const methods = [
	'addCookie',
	'clearCookies',
	'close',
	'deleteCookie',
	'evaluateJavaScript',
	'injectJs',
	'openUrl',
	'property',
	'reload',
	'render',
	'renderBase64',
	'sendEvent',
	'setContent',
	'setting',
	'stop',
	'switchToFrame',
    'evaluate',
    'switchToMainFrame'
];

asyncMethods.forEach(function (method) {
	Page.prototype[method] = function () {
		return this.invokeAsyncFunction.apply(this, [method].concat([].slice.call(arguments)));
	};
});

methods.forEach(function (method) {
	Page.prototype[method] = function () {
        return this.invokeFunction.apply(this, [method].concat([].slice.call(arguments)));
    };
});
```

What do you think?

**Comment by amir20** [comment]
I like most of this. One thing that worries me is the use of `property` to add a new function. I feel it should be something more clear, ie. `page.defineFunction`.

**Comment by amir20** [comment]
@elias-winberg did you want to create a pull request with these changes and I can merge?

**Comment by elias-winberg** [comment]
@amir20 I'll give it a try!

**Comment by amir20** [comment]
I believe in you! :) 👍

**Comment by elias-winberg** [comment]
Which style do you prefer?

```js
#.invokeFunction(functionName, [arg1, arg2, arg3])
```

```js
#.invokeFunction(functionName, arg1, arg2, arg3)
```

**Comment by amir20** [comment]
If it's external, then the second one because it is more intuitive. I know internally I have done `[...]` because I have to pass arrays around.

**Comment by amir20** [comment]
Fixed and 2.1.8 released.
